### PR TITLE
[VecOps][Wip] Introduce the RStrongBool type for representing the results of logical operators

### PR DESCRIPTION
**Pull request by dpiparo**
Before, the boolean type considered by the vecops was int.

### CODE REVIEW

File: math/vecops/inc/ROOT/RStrongBool.hxx [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+}
+private:
+  int fVal;
+};
+
+} // End NS VecOps
+} // End NS ROOT
+
+// this is inlined to workaround missing symbols errors
+inline std::ostream& operator<< (std::ostream& stream, ROOT::VecOps::RStrongBool sb)
```
Comment (bluehood) [inline math/vecops/inc/ROOT/RStrongBool.hxx]: this operator is breaking some tests

File: math/vecops/inc/ROOT/RStrongBool.hxx [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  template<typename T>
+  RStrongBool(T val) = delete;
+  RStrongBool(bool val);
+  // this is inlined to workaround missing symbols errors
+  explicit operator bool() const
+{
+	return bool(fVal);
+}
+private:
+  int fVal;
```
Comment (bluehood) [inline math/vecops/inc/ROOT/RStrongBool.hxx]: why not bool?

File: math/vecops/inc/ROOT/RVec.hxx [hunk 3]
```diff
@@ -642,21 +642,19 @@ RVec<T> Filter(const RVec<T> &v, F &&f)
 }
 
 /// Return true if any of the elements equates to true, return false otherwise.
-template <typename T>
-auto Any(const RVec<T> &v) -> decltype(v[0] == true)
+inline bool Any(const RVec<RStrongBool> &v)
```
Comment (bluehood) [inline math/vecops/inc/ROOT/RVec.hxx]: it's a bit weird that the following shouldn't compile:
```c++
RVec<bool> vals{true,false};
Any(vals);
```

Maybe `RVec<bool>` should be implicitly convertible to `RVec<RStrongBool>`? Or is it too messy?

File: math/vecops/inc/ROOT/RVec.hxx [hunk 4]
```diff
@@ -673,22 +671,41 @@ template <class T>
...
+         os << Print_t(v[i]) << ", ";
+      }
+      os << Print_t(v[size - 1]);
+   }
+   os << " }";
+   return os;
+}
+
+// Workaround for cling.
+// If one tries to instantiate at the prompt or in a macro an RVec<RStrongBool>
```
Comment (bluehood) [inline math/vecops/inc/ROOT/RVec.hxx]: does this mean that users cannot write `RVec<RStrongBool>` at the prompt?

File: math/vecops/src/RStrongBool.cxx [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+#include "ROOT/RStrongBool.hxx"
+
+#include <string>
+
+namespace ROOT {
+namespace VecOps {
+
+RStrongBool::RStrongBool() : fVal(false){};
+RStrongBool::~RStrongBool() = default;
+RStrongBool::RStrongBool(bool val) : fVal{val}
```
Comment (bluehood) [inline math/vecops/src/RStrongBool.cxx]: why are these simple ctors and dtor out of line?

File: math/vecops/src/RStrongBool.cxx [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+ *************************************************************************/
+
+#include "ROOT/RStrongBool.hxx"
+
+#include <string>
+
+namespace ROOT {
+namespace VecOps {
+
+RStrongBool::RStrongBool() : fVal(false){};
```
Comment (bluehood) [inline math/vecops/src/RStrongBool.cxx]: why having a default value at all? would the removal of a default ctor make it awkward to use?

File: math/vecops/inc/ROOT/RStrongBool.hxx [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  RStrongBool();
+  ~RStrongBool();
+  template<typename T>
+  RStrongBool(T val) = delete;
+  RStrongBool(bool val);
+  // this is inlined to workaround missing symbols errors
+  explicit operator bool() const
+{
+	return bool(fVal);
+}
```
Comment (bluehood) [inline math/vecops/inc/ROOT/RStrongBool.hxx]: why must the cast be explicit? one would reasonably expect `RStrongBool` to be usable in place of a `bool`

**Comment by bluehood** [comment]
@amadio The end goal is to allow index manipulations with TVec.
I don't know if there are other reasons, but one is to distinguish indexing with a mask `v[{true, false}]` from indexing with a list of indeces (currently unsupported, but possibly very useful) `v[{1,2}]`.
The second could also be done with something like `Take(v, {1,2})`, now that I think about it.

**Comment by amadio** [comment]
What does `v[{true, false}]` mean? I'm still not convinced that this is necessary.
Also, why do you need `v[{1,2}]` instead of `{v[1], v[2]}`?

**Comment by amadio** [comment]
You could also have something like
```cpp
template <typename T>
TVec<T> MakeVec(TVec<T> const &v, std::initializer_list<int> &idx) { ... }
```

**Comment by amadio** [comment]
Note that `v[v >0]` is meant to return a reference to a masked vector so that one can do `v[v > 0] = 2;`, so maybe that's what we need to fix. What you mention above looks like targeted at creating copies of parts of a vector, which may lead to surprising behavior from user's point of view.

**Comment by bluehood** [comment]
@amadio note that the current implementation of `v[v > 2]` returns a new vector, not a vector of references (not straightforward to have a vector of references to non-contiguous storage in c++) but it's not a bug, it was designed like that. I agree that this makes `v[v>2] = 2` very misleading (compiles but assigns to a temporary, I guess) -- but that's a different topic..?

**Comment by pcanal** [comment]
>  I agree that this makes v[v>2] = 2 very misleading (compiles but assigns to a temporary, I guess)

It would probably be best if this is made illegal at compile time ... By having either the return value of v[..] to be const.

**Comment by dpiparo** [comment]
Not needed anymore!!

**Pull request closed without merging**